jib-gradle-plugin/src/main/java/com/google/cloud/tools/jib/gradle/FilesTask.java

```diff
@@ -0,0 +1,174 @@
+      System.out.println(projectPath.resolve(Settings.DEFAULT_SETTINGS_FILE));
+    }
+
+    // Print gradle.properties
+    if (Files.exists(projectPath.resolve("gradle.properties"))) {
+      System.out.println(projectPath.resolve("gradle.properties"));
+    }
+  }
+
+  /**
+   * Prints build files, sources, and resources associated with a project.
+   *
+   * @param project the project
+   */
+  private static void printProjectFiles(Project project) {
+    JavaPluginConvention javaConvention =
+        project.getConvention().getPlugin(JavaPluginConvention.class);
+    SourceSet mainSourceSet =
+        javaConvention.getSourceSets().findByName(SourceSet.MAIN_SOURCE_SET_NAME);
+    if (mainSourceSet == null) {
+      return;
+    }
+
+    // Print build config, settings, etc.
+    printGradleFiles(project);
+
+    // Print sources + resources
+    mainSourceSet
+        .getAllSource()
+        .getSourceDirectories()
+        .forEach(
+            sourceDirectory -> {
+              if (sourceDirectory.exists()) {
+                System.out.println(sourceDirectory);
+              }
+            });
+  }
+
+  /**
+   * Recursive function for printing out a project's artifacts. Calls itself when it encounters a
+   * project dependency.
+   *
+   * @param project the project to list the artifacts for
+   * @param projectDependenciesResult the set of project dependencies encountered. When a project
+   *     dependency is encountered, it is added to this set.
+   */
+  private static void findProjectDependencies(
+      Project project, Set<ProjectDependency> projectDependenciesResult) {
+    for (Configuration configuration :
+        project.getConfigurations().getByName("runtime").getHierarchy()) {
+      for (Dependency dependency : configuration.getDependencies()) {
+        if (dependency instanceof ProjectDependency) {
+          projectDependenciesResult.add((ProjectDependency) dependency);
+          findProjectDependencies(
+              ((ProjectDependency) dependency).getDependencyProject(), projectDependenciesResult);
+        }
+      }
+    }
+  }
+
+  @Nullable private JibExtension jibExtension;
+
+  public FilesTask setJibExtension(JibExtension jibExtension) {
+    this.jibExtension = jibExtension;
+    return this;
+  }
+
+  @TaskAction
+  public void listFiles() {
+    Preconditions.checkNotNull(jibExtension);
+    Project project = getProject();
+
+    // If this is not the root project, print the root project's build.gradle and settings.gradle
+    if (project != project.getRootProject()) {
+      printGradleFiles(project.getRootProject());
+    }
+
+    printProjectFiles(project);
+
+    // Print extra layer
+    if (Files.exists(jibExtension.getExtraDirectoryPath())) {
+      System.out.println(jibExtension.getExtraDirectoryPath());
+    }
+
+    // Find project dependencies
+    Set<ProjectDependency> projectDependencies = new HashSet<>();
+    findProjectDependencies(project, projectDependencies);
```

Comment:
nit: Usually prefer a more functional style: `Set<ProjectDependency> projectDependencies = findProjectDependencies(project);`

Reply:
Perhaps in another PR we can flatten it out using a non-recursive BFS or DFS (So you're not constantly creating and merging sets).

Reply:
Filed #1033